File: src/main/java/org/cimsbioko/repository/CursorConvert.java
package org.cimsbioko.repository;

import android.database.Cursor;

import java.util.ArrayList;
import java.util.List;

public class CursorConvert {

    public static <X> List<X> list(Cursor cursor, CursorConverter<X> converter) {
        List<X> result = new ArrayList<>();
        if (cursor != null) {
            try {
                while (cursor.moveToNext()) {
                    result.add(converter.convert(cursor));
                }
            } finally {
                cursor.close();
            }
        }
        return result;
    }

    public static <Y> Y one(Cursor cursor, CursorConverter<Y> converter) {
        if (cursor != null) {
            try {
                if (cursor.moveToFirst()) {
                    return converter.convert(cursor);
                }
            } finally {
                cursor.close();
            }
        }
        return null;
    }

}

File: src/main/java/org/cimsbioko/data/CursorConvert.kt
package org.cimsbioko.data

import android.database.Cursor

internal object CursorConvert {

    @JvmStatic
    fun <X> list(cursor: Cursor?, converter: CursorConverter<X>): List<X> =
            ArrayList<X>().apply {
                cursor?.use {
                    while (it.moveToNext()) {
                        add(converter.convert(it))
                    }
                }
            }

    @JvmStatic
    fun <Y> one(cursor: Cursor?, converter: CursorConverter<Y>): Y? =
            cursor?.use { if (it.moveToFirst()) converter.convert(it) else null }

    @JvmStatic
    fun extractString(cursor: Cursor, columnName: String): String? =
            cursor.getColumnIndex(columnName).let { if (it >= 0) cursor.getString(it) else null }

}